src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs

```diff
@@ -2,49 +2,58 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Linq;
 using System.Runtime.InteropServices;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
 using Microsoft.CodeAnalysis.Shared.TestHooks;
 using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CSharp.RuntimeBinder;
+using Microsoft.VisualStudio.Editor;
 using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
 using Microsoft.VisualStudio.TextManager.Interop;
 using Roslyn.Utilities;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.Snippets
 {
+    /// <summary>
+    /// This service is created on the UI thread during package initialization, but it must not
+    /// block the initialization process. If the expansion manager is an IExpansionManager,
+    /// then we can use the asynchronous population mechanism it provides. Otherwise, getting
+    /// snippet information from the <see cref="IVsExpansionManager"/> must be done synchronously
+    /// through on the UI thread, which we do after package initialization at a lower priority.
+    /// </summary>
+    /// <remarks>
+    /// IExpansionManager was introduced in Visual Studio 2015 Update 1, but
+    /// will be enabled by default for the first time in Visual Studio 2015 Update 2. However,
+    /// the platform still supports returning the <see cref="IVsExpansionManager"/> if a major
+    /// problem in the IExpansionManager is discovered, so we must continue 
+    /// supporting the fallback.
+    /// </remarks>
     internal abstract class AbstractSnippetInfoService : ForegroundThreadAffinitizedObject, ISnippetInfoService, IVsExpansionEvents
     {
         private readonly Guid _languageGuidForSnippets;
         private readonly IVsExpansionManager _expansionManager;
 
         /// <summary>
-        /// This service is created on the UI thread during package initialization, but it must not
-        /// block the initialization process. Getting snippet information from the <see cref="IVsExpansionManager"/>
-        /// must be done on the UI thread, so do this work in a task that will run on the UI thread
-        /// with lower priority.
+        /// Initialize these to empty values. When returning from <see cref="GetSnippetsIfAvailable "/> 
+        /// and <see cref="SnippetShortcutExists_NonBlocking"/>, we return the current set of known 
+        /// snippets rather than waiting for initial results.
         /// </summary>
-        protected readonly Task InitialCachePopulationTask;
-
-        protected object cacheGuard = new object();
-
-        // Initialize these to empty values. When returning from GetSnippetsIfAvailable and
-        // SnippetShortcutExists_NonBlocking, we can return without checking the status
-        // of InitialCachePopulationTask.
         protected IList<SnippetInfo> snippets = SpecializedCollections.EmptyList<SnippetInfo>();
```

Comment:
Consider: ImmutableArray or some other type to clarify your threading assumptions.